src/mem/cache/tags/tagged_entry.hh

```diff
@@ -102,38 +191,55 @@ class TaggedEntry : public CacheEntry
             isSecure(), isValid(), ReplaceableEntry::print());
     }
 
-    bool
-    matchTag(const Addr tag) const override
-    {
-        panic("Need is_secure arg");
-        return false;
-    }
-
-    void
-    insert(const Addr tag) override
-    {
-        panic("Need is_secure arg");
-        return;
-    }
   protected:
+    /**
+     * Set tag associated to this block.
+     *
+     * @param tag The tag value.
+     */
+    virtual void setTag(Addr tag) { _tag = tag; }
+
     /** Set secure bit. */
     virtual void setSecure() { _secure = true; }
 
+    /** Clear secure bit. Should be only used by the invalidation function. */
+    void clearSecure() { _secure = false; }
+
+    /** Set valid bit. The block must be invalid beforehand. */
+    virtual void
+    setValid()
+    {
+        assert(!isValid());
+        _valid = true;
+    }
+
+    /** Callback used to extract the tag from the entry */
+    TagExtractor extractTag;
+
   private:
+    /**
+     * Valid bit. The contents of this entry are only valid if this bit is set.
+     * @sa invalidate()
+     * @sa insert()
+     */
+    bool _valid;
+
     /**
      * Secure bit. Marks whether this entry refers to an address in the secure
      * memory space. Must always be modified along with the tag.
      */
     bool _secure;
 
-    /** Clear secure bit. Should be only used by the invalidation function. */
-    void clearSecure() { _secure = false; }
-
-    /** Do not use API without is_secure flag. */
-    using CacheEntry::matchTag;
-    using CacheEntry::insert;
+    /** The entry's tag. */
+    Addr _tag;
 };
 
+static constexpr auto
+genTagExtractor(TaggedIndexingPolicy *ip)
```

Comment:
Can you add a couple of lines of documentation here? I think this means "generic" tag extractor, but I'm not certain.

Also, this is magic c++ syntax (at least to me). Could you add a line explaining what the syntax is doing?

Reply:
It actually stands for "generate tag extractor" as it returns a lambda. I will document this better